cirq-core/cirq/value/classical_data.py

```diff
@@ -0,0 +1,278 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import enum
+from typing import Dict, Mapping, Sequence, Tuple, TYPE_CHECKING, TypeVar
+
+from cirq.value import digits, value_equality_attr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class MeasurementType(enum.IntEnum):
+    MEASUREMENT = 1
+    CHANNEL = 2
+
+    def __repr__(self):
+        return f'cirq.{str(self)}'
+
+
+TSelf = TypeVar('TSelf', bound='ClassicalDataStoreReader')
+
+
+class ClassicalDataStoreReader(abc.ABC):
+    @abc.abstractmethod
+    def keys(self) -> Tuple['cirq.MeasurementKey', ...]:
+        """Gets the measurement keys in the order they were stored."""
+
+    @abc.abstractmethod
+    def get_int(self, key: 'cirq.MeasurementKey') -> int:
+        """Gets the integer corresponding to the measurement.
+
+        Args:
+            key: The measurement key.
+
+        Raises:
+            KeyError: If the key has not been used.
+        """
+
+    @abc.abstractmethod
+    def get_digits(self, key: 'cirq.MeasurementKey') -> Tuple[int, ...]:
+        """Gets the digits of the measurement.
+
+        Args:
+            key: The measurement key.
+
+        Raises:
+            KeyError: If the key has not been used.
+        """
+
+    @abc.abstractmethod
+    def copy(self: TSelf) -> TSelf:
+        """Creates a copy of the object."""
+
+
+class ClassicalDataStore(ClassicalDataStoreReader, abc.ABC):
+    @abc.abstractmethod
+    def record_measurement(
+        self, key: 'cirq.MeasurementKey', measurement: Sequence[int], qubits: Sequence['cirq.Qid']
+    ):
+        """Records a measurement.
+
+        Args:
+            key: The measurement key to hold the measurement.
+            measurement: The measurement result.
+            qubits: The qubits that were measured.
+
+        Raises:
+            ValueError: If the measurement shape does not match the qubits
+                measured or if the measurement key was already used.
+        """
+
+    @abc.abstractmethod
+    def record_channel_measurement(self, key: 'cirq.MeasurementKey', measurement: int):
+        """Records a channel measurement.
+
+        Args:
+            key: The measurement key to hold the measurement.
+            measurement: The measurement result.
```

Comment:
Maybe the Kraus probabilities? I didn't think the qubits were relevant here so didn't include them, figuring it would create confusion that the value here could be larger than the product of qid dimensions. I figured if we determined that we do need something, it's easy to add later?